util/logtools/make_timeline.ts

```diff
@@ -0,0 +1,512 @@
+  console.error(errString);
+  timelineParse.parser.printHelp();
+  process.exit(-1);
+};
+
+// TODO: Revisit this logic when we re-add FFLogs support.
+if (args.file === null)
+  printHelpAndExit('Error: Must specify -f\n');
+if (!args.file?.includes('.log'))
+  printHelpAndExit('Error: Must specify an FFXIV ACT log file, as log.log\n');
+let numExclusiveArgs = 0;
+for (const opt of ['search_fights', 'search_zones', 'fight_regex', 'zone_regex']) {
+  if (args[opt] !== null)
+    numExclusiveArgs++;
+}
+if (numExclusiveArgs !== 1)
+  printHelpAndExit('Error: Must specify exactly one of -lf, -lz, or -fr\n');
+if (args['fight_regex'] === '-1')
+  printHelpAndExit('Error: -fr must specify a regex\n');
+if (args['zone_regex'] === '-1')
+  printHelpAndExit('Error: -zr must specify a regex\n');
+
+if (args.file && numExclusiveArgs !== 1 && !(args.start && args.end)) {
+  printHelpAndExit(
+    'Error: -s and -e must be specified if not using one of -lf, -lz, -fr, or -zr\ ',
+  );
+}
+if (args.start && !(typeof args.start === 'string' && timeStampValidate(args.start)))
+  printHelpAndExit('Error: Timestamp must be entered exactly as HH:MM:SS.DDD, as 12:34:56:.789');
+if (args.end && !(typeof args.end === 'string' && timeStampValidate(args.end)))
+  printHelpAndExit('Error: Timestamp must be entered exactly as HH:MM:SS.DDD, as 12:34:56:.789');
+
+const makeCollectorFromPrepass = async (fileName: string) => {
+  const collector = new EncounterCollector();
+  const lineReader = readline.createInterface({
+    input: fs.createReadStream(fileName),
+  });
+  for await (const line of lineReader) {
+    // TODO: this could be more efficient if it stopped when it found the requested encounter.
+    collector.process(line);
+  }
+  return collector;
+};
+
+const parseNameToggleToEntry = (matches: NetMatches['NameToggle']): TimelineEntry => {
+  const targetable = matches.toggle === '01';
+  const entry: TimelineEntry = {
+    time: matches.timestamp,
+    combatant: matches.name,
+    lineType: 'nameToggle',
+    targetable: targetable,
+  };
+  return entry;
+};
+
+const parseAbilityToEntry = (matches: NetMatches['Ability']): TimelineEntry => {
+  let abilityName = matches.ability;
+  if (abilityName.toLowerCase().includes('unknown_'))
+    abilityName = '--sync--';
+  const entry: TimelineEntry = {
+    time: matches.timestamp,
+    combatant: matches.source,
+    abilityId: matches.id,
+    abilityName: abilityName,
+    lineType: 'ability',
+  };
+  return entry;
+};
+
+const extractRawLines = async (fileName: string, start: string, end: string): Promise<string[]> => {
+  const lines: string[] = [];
+  const file = readline.createInterface({
+    input: fs.createReadStream(fileName),
+  });
+  let started = false;
+  for await (const line of file) {
+    // This will fail on lines with 3-digit identifiers,
+    // but that's okay because those will never be start lines.
+    const lineTimeStamp = line.substring(14, 26);
```

Comment:
For this logic (and elsewhere in the file), RaidEmulator's ParseLine could be leveraged instead (and maybe it's time for that functionality to move out of the RaidEmulator subset of code and into `resources` or something).

Reply:
Yes, you're right. I haven't really looked at the internals of the emulator much, and this was intended as a foundational mostly 1-1 conversion, but looking through `EmulatorCommon` there's a lot of time handling that could be brought into resources and removed from several other places. I think it might be better to do that as a follow-up after landing this (and maybe your `test` implementation as well.)